pkg/strategy/streaming.go

```diff
@@ -15,84 +17,235 @@ import (
+
+	// we provision a Kafka when no brokers have been set, or, when we are not in the first run,
+	// when we know we've been the ones placing the broker information in the configuration
+	if (!pfound && !cfound) || provisioned {
+		manifest = autoProvisionKafka(jaeger, manifest)
 	}
 
 	// add the deployments
-	c.deployments = []appsv1.Deployment{*collector.Get(), *inject.Sidecar(jaeger, inject.OAuthProxy(jaeger, query.Get()))}
+	manifest.deployments = []appsv1.Deployment{*collector.Get(), *inject.Sidecar(jaeger, inject.OAuthProxy(jaeger, query.Get()))}
 
 	if d := ingester.Get(); d != nil {
-		c.deployments = append(c.deployments, *d)
+		manifest.deployments = append(manifest.deployments, *d)
 	}
 
 	// add the daemonsets
 	if ds := agent.Get(); ds != nil {
-		c.daemonSets = []appsv1.DaemonSet{*ds}
+		manifest.daemonSets = []appsv1.DaemonSet{*ds}
 	}
 
 	// add the services
 	for _, svc := range collector.Services() {
-		c.services = append(c.services, *svc)
+		manifest.services = append(manifest.services, *svc)
 	}
 
 	for _, svc := range query.Services() {
-		c.services = append(c.services, *svc)
+		manifest.services = append(manifest.services, *svc)
 	}
 
 	// add the routes/ingresses
 	if viper.GetString("platform") == v1.FlagPlatformOpenShift {
 		if q := route.NewQueryRoute(jaeger).Get(); nil != q {
-			c.routes = append(c.routes, *q)
+			manifest.routes = append(manifest.routes, *q)
 		}
 	} else {
 		if q := ingress.NewQueryIngress(jaeger).Get(); nil != q {
-			c.ingresses = append(c.ingresses, *q)
+			manifest.ingresses = append(manifest.ingresses, *q)
 		}
 	}
 
 	if isBoolTrue(jaeger.Spec.Storage.Dependencies.Enabled) {
 		if cronjob.SupportedStorage(jaeger.Spec.Storage.Type) {
-			c.cronJobs = append(c.cronJobs, *cronjob.CreateSparkDependencies(jaeger))
+			manifest.cronJobs = append(manifest.cronJobs, *cronjob.CreateSparkDependencies(jaeger))
 		} else {
 			jaeger.Logger().WithField("type", jaeger.Spec.Storage.Type).Warn("Skipping spark dependencies job due to unsupported storage.")
 		}
 	}
 
 	if isBoolTrue(jaeger.Spec.Storage.EsIndexCleaner.Enabled) {
 		if strings.EqualFold(jaeger.Spec.Storage.Type, "elasticsearch") {
-			c.cronJobs = append(c.cronJobs, *cronjob.CreateEsIndexCleaner(jaeger))
+			manifest.cronJobs = append(manifest.cronJobs, *cronjob.CreateEsIndexCleaner(jaeger))
 		} else {
 			jaeger.Logger().WithField("type", jaeger.Spec.Storage.Type).Warn("Skipping Elasticsearch index cleaner job due to unsupported storage.")
 		}
 	}
 
-	c.dependencies = storage.Dependencies(jaeger)
+	manifest.dependencies = storage.Dependencies(jaeger)
+
+	return manifest
+}
+
+func autoProvisionKafka(jaeger *v1.Jaeger, manifest S) S {
+	if jaeger.Annotations == nil {
+		jaeger.Annotations = map[string]string{}
+	}
+	// mark that we auto provisioned a kafka for this instance
+	jaeger.Annotations[v1.AnnotationProvisionedKafkaKey] = v1.AnnotationProvisionedKafkaValue
```

Comment:
I see the purpose here - but if required, then need to change the key so that it is kafka specific. This also implies that the `newStreamingStrategy` is called multiple times? is it per reconciliation cycle?

Reply:
But then why create new `kafka` and `kafkauser` if already provisioned? Sorry possibly lack of understanding of the mechanism.

Reply:
Perhaps the comment I left for the previous item clarified this already, but if not, let me know.